src/Mvc/Mvc.Core/src/Builder/ControllerEndpointRouteBuilderExtensions.cs

```diff
@@ -447,6 +448,44 @@ public static ControllerActionEndpointConventionBuilder MapDefaultControllerRout
             return builder;
         }
 
+        /// <summary>
+        /// Adds a specialized <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that will
+        /// attempt to select a controller action using the route values produced by <typeparamref name="TTransformer"/>.
+        /// </summary>
+        /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/> to add the route to.</param>
+        /// <param name="pattern">The URL pattern of the route.</param>
+        /// <typeparam name="TTransformer">The type of a <see cref="DynamicRouteValueTransformer"/>.</typeparam>
+        /// <remarks>
+        /// <para>
+        /// This method allows the registration of a <see cref="RouteEndpoint"/> and <see cref="DynamicRouteValueTransformer"/>
+        /// that combine to dynamically select a controller action using custom logic.
+        /// </para>
+        /// <para>
+        /// The instance of <typeparamref name="TTransformer"/> will be retrieved from the dependency injection container.
+        /// Register <typeparamref name="TTransformer"/> with the desired service lifetime in <c>ConfigureServices</c>.
+        /// </para>
+        /// </remarks>
+        public static void MapDynamicControllerRoute<TTransformer>(this IEndpointRouteBuilder endpoints, string pattern)
+            where TTransformer : DynamicRouteValueTransformer
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            EnsureControllerServices(endpoints);
+
+            // Called for side-effect to make sure that the data source is registered.
+            GetOrCreateDataSource(endpoints).CreateInertEndpoints = true;
+
+            // Maps a dynamic controller endpoint with an empty delegate. This is OK because
+            // we don't expect the delegate to run.
+            endpoints.Map(pattern, context => Task.CompletedTask).Add(b =>
```

Comment:
Should the delegate throw so we when stuff's broken?

Reply:
Sure.

Reply:
I just want to take a moment, and mention one thing. 

I agree that failing loudly in this case is better than silently. However, usually when this happens, the next piece of feedback is to provide an descriptive and meaningful error message for the code that we don't expect to ever run.